cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -344,4 +349,17 @@ public void removeStreamTags(HttpRequest request, HttpResponder responder,
   public void recordRun(HttpRequest request, HttpResponder responder) {
     responder.sendString(HttpResponseStatus.OK, "Metadata recorded successfully");
   }
+
+  // *** Search endpoints ***
+
+  @GET
+  @Path("/namespaces/{namespace-id}/metadata/search")
+  public void searchMetadata(HttpRequest request, HttpResponder responder,
+                             @PathParam("namespace-id") String namespaceId,
+                             @QueryParam("query") String searchQuery,
+                             @QueryParam("target") MetadataSearchTargetType target) throws Exception {
+    Iterable<MetadataSearchResultRecord> results = metadataAdmin.searchMetadata(searchQuery, target);
+
+    responder.sendJson(HttpResponseStatus.OK, Sets.newHashSet(results));
```

Comment:
Set since the returned data is the distinct objects inverted index by the value or key=value.
The dataset return list of actual records coming which could have duplicate target objects.

For example. If you search for value of "blah" and someone had 2 key-value pairs like key1=blah and key2=blah then the Dataset will return 2 records.